File: SBUS-Android/java/tools/Persist.java
import java.util.*;
import java.util.concurrent.locks.*;
import java.io.*;

class Persistence
{
	ArrayList<String> cpts, procs;
	String sbus_dir;
	
	Persistence()
	{
		String filename;

		procs = null;

		// Read ~/.sbus/autorun
		filename = Log.get_sbus_dir() + "/autorun";
		cpts = Utils.readlinefile(filename, true);
		if(cpts == null)
		{
			Log.warning("Warning: Could not open " + filename +
					"; no persistent components list");
			cpts = new ArrayList<String>();
		}
	}
	
	void start(String cmd)
	{
		String[] args;
		Process p;
		Runtime runtime = Runtime.getRuntime();

		args = cmd.split("[\t ]+");
		try
		{
			p = runtime.exec(args);
		}
		catch(IOException e)
		{
			Log.warning("Warning: could not start persistent component -\n" +
					cmd);
		}
	}

	void start_all()
	{
		procs = running_procs();

		for(int i = 0; i < cpts.size(); i++)
		{
			checkquotes(cpts.get(i));
			System.out.println("Persistent cpt " + cpts.get(i) +
					(is_running(cpts.get(i)) ? " is running" : " is not running"));
			if(!is_running(cpts.get(i)))
				start(cpts.get(i));
		}
	}
	
	boolean is_persistent(String cmd)
	{
		for(int i = 0; i < cpts.size(); i++)
		{
			if(cpts.get(i).equals(cmd))
				return true;
		}
		return false;
	}

	private ArrayList<String> running_procs()
	{
		File dir;
		String[] names;
		BufferedReader in;
		String pathname, s;
		ArrayList<String> results = new ArrayList<String>();

		// Access /proc/<number>/cmdline
		dir = new File("/proc");
		if(!dir.isDirectory())
		{
			Log.warning("Cannot open /proc directory; " +
					"disabling running process check");
			return results;
		}
		names = dir.list();
		for(String filename: names)
		{
			if(!is_an_integer(filename))
				continue;
			pathname = "/proc/" + filename + "/cmdline";
			try
			{
				in = new BufferedReader(new FileReader(pathname));
				s = in.readLine(); // Only need to read one line
				if(s != null)
				{
					StringBuilder sb = new StringBuilder(s);
					for(int i = 0; i < sb.length(); i++)
					{
						if(sb.charAt(i) == '\0')
							sb.setCharAt(i, ' ');
					}
					results.add(sb.toString());
				}
				in.close();
			}
			catch(IOException e)
			{
				Log.warning("Warning: could not open " + pathname);
				continue;
			}
		}
		return results;
	}
	
	private boolean is_running(String proc)
	{
		for(int i = 0; i < procs.size(); i++)
		{
			if(procs.get(i).equals(proc))
				return true;
		}
		return false;
	}

	void hup(Endpoint ep, ArrayList<Image> live, Lock live_mutex,
			Endpoint terminate_ep)
	{
		Image img;
		String filename;
		ProtoMessage dummy;
		
		dummy = ep.rcv(); // Discard dummy

		System.out.println("Received hangup instruction: " +
				"re-checking persistent components list\n");

		// Refresh list of persistent components:
		filename = sbus_dir + "/autorun";
		cpts = Utils.readlinefile(filename, true);
		if(cpts == null)
		{
			Log.warning("Warning: Could not open " + filename +
					"; no persistent components list");
			cpts = new ArrayList<String>();
		}

		// Scan for any which are no longer persistent, and stop them:
		live_mutex.lock();
		for(int i = 0; i < live.size(); i++)
		{
			img = live.get(i);
			if(img.local && img.persistent)
			{
				if(!is_persistent(img.state.extract_txt("cmdline")))
				{
					// Stop it:
					System.out.println("Terminating previously persistent " +
							"component " + img.address);
					img.persistent = false;
					if(terminate_ep.map(img.address, "terminate") == null)
					{
						System.out.println("Failed to terminate " + img.address);
						continue;
					}
					terminate_ep.emit(null);
					terminate_ep.unmap();
				}
			}
		}
		live_mutex.unlock();

		// Start any which are not running:
		start_all();
	}
	
	// Utilities:
			
	private boolean is_an_integer(String s)
	{
		int pos, len = s.length();
		if(len == 0)
			return false;
		for(pos = 0; pos < len; pos++)
			if(s.charAt(pos) < '0' || s.charAt(pos) > '9')
				return false;
		if(pos > 20) // Integer getting too long for buffers
			return false;
		return true;
	}
	
	private void checkquotes(String cmd)
	{
		if(cmd.contains("\"") || cmd.contains("'"))
		{
			Log.error("Quotation marks not currently allowed in persistent " +
					"component command lines");
		}
	}
};

File: SBUS-Android/tools/persist.cpp
// persist.cpp - DMI - 4-8-2008

#include <stdio.h>
#include <stdlib.h>
#include <dirent.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <pthread.h>

#include <sys/types.h>
#include <sys/stat.h>

#include "../library/sbus.h"

#include "rdc.h"

typedef char *charptr;

persistence::persistence()
{
	char *filename;

	procs = NULL;
		
	// Read ~/.sbus/autorun
	sbus_dir = get_sbus_dir();
	filename = new char[strlen(sbus_dir) + 20];	
	sprintf(filename, "%s/autorun", sbus_dir);
	cpts = new linefile(filename, 1);
	if(!cpts->valid())
	{
		warning("Warning: Could not open %s; no persistent components list",
				filename);
	}
	delete[] filename;
}

void persistence::hup(sendpoint *ep, imagevector *live,
		pthread_mutex_t *live_mutex, sendpoint *terminate_ep)
{
	char *filename;
	image *img;
	
	smessage *dummy;
	dummy = ep->rcv();
	delete dummy;
	
	printf("Received hangup instruction: re-checking persistent components "
			"list\n");
	
	// Refresh list of persistent components:
	delete cpts;
	filename = new char[strlen(sbus_dir) + 20];	
	sprintf(filename, "%s/autorun", sbus_dir);
	cpts = new linefile(filename, 1);
	if(!cpts->valid())
	{
		warning("Warning: Could not open %s; no persistent components list",
				filename);
	}
	delete[] filename;
	
	// Scan for any which are no longer persistent, and stop them:
	pthread_mutex_lock(live_mutex);
	for(int i = 0; i < live->count(); i++)
	{
		img = live->item(i);
		if(img->local && img->persistent)
		{
			if(!is_persistent(img->state->extract_txt("cmdline")))
			{
				// Stop it:
				printf("Terminating previously persistent component %s\n",
						img->address);
				img->persistent = 0;
				if(!terminate_ep->map(img->address, "terminate"))
				{
					printf("Failed to terminate %s\n", img->address);
					continue;
				}
				terminate_ep->emit(NULL);
				terminate_ep->unmap();
			}
		}
	}
	pthread_mutex_unlock(live_mutex);
	
	// Start any which are not running:
	start_all();
}

void persistence::start_all()
{
	if(procs != NULL)
		delete procs;
	procs = running_procs();

	for(int i = 0; i < cpts->count(); i++)
	{
		checkquotes(cpts->item(i));
		printf("Persistent cpt %s %s\n", cpts->item(i),
				(is_running(cpts->item(i)) ? "is running" : "is not running"));
		if(!is_running(cpts->item(i)))
			start(cpts->item(i));
	}
}

void persistence::checkquotes(const char *cmd)
{
	while(*cmd != '\0')
	{
		if(*cmd == '"' || *cmd == '\'')
		{
			error("Quotation marks not currently allowed in persistent component"
					" command lines");
		}
		cmd++;
	}
}

svector *persistence::split(const char *s)
{
	svector *v = new svector();
	const char *start;

	while(*s == ' ' || *s == '\t')
		s++;
	start = s;
	while(1)
	{
		if(*s == '\0')
		{
			if(start != s)
				v->add(start, s - start);
			break;
		}
		if(*s == ' ' || *s == '\t')
		{
			v->add(start, s - start);
			while(*s == ' ' || *s == '\t')
				s++;
			start = s;
		}
		else
			s++;
	}
	return v;
}

void persistence::detach()
{
	// Detach from parent
	int fd;
	
	close(0);
	close(1);
	close(2);
	
	fd = open("/dev/null", O_RDWR); // stdin
	dup(fd); // stdout
	dup(fd); // stderr
	
	if(fork() > 0)
		exit(0);
}

void persistence::start(const char *cmd)
{
	int pid;

	// Create new process:
	pid = fork();
	if(pid != 0)
		return;

	detach();
	
	int ret;
	char *filename, **argv;
	svector *v;
	int args;

	v = split(cmd);
	args = v->count();
	filename = v->item(0);
	argv = new charptr[args + 1];
	for(int i = 0; i < args; i++)
		argv[i] = v->item(i);
	argv[args] = NULL;
	
	ret = execvp(filename, argv);
	delete v;
	printf("Warning: could not start persistent component -\n%s\n", cmd);
	printf("%s\n", strerror(errno));
	exit(0);
}

int persistence::is_persistent(const char *cmd)
{
	for(int i = 0; i < cpts->count(); i++)
	{
		if(!strcmp(cpts->item(i), cmd))
			return 1;
	}
	return 0;
}

int persistence::is_running(const char *proc)
{
	for(int i = 0; i < procs->count(); i++)
	{
		if(!strcmp(procs->item(i), proc))
			return 1;
	}
	return 0;
}

persistence::~persistence()
{
	delete[] sbus_dir;
	if(cpts != NULL)
		delete cpts;
	if(procs != NULL)
		delete procs;
}

int persistence::is_an_integer(const char *name)
{
	int pos = 0;
	
	if(name[pos] < '0' || name[pos] > '9')
		return 0;
	pos++;
	
	while(1)
	{
		if(name[pos] == '\0')
			break;
		if(name[pos] < '0' || name[pos] > '9')
			return 0;
		pos++;
		if(pos > 20) // Integer getting too long for buffers
			return 0;
	}
	return 1;
}

svector *persistence::running_procs()
{
	struct dirent *de;
	DIR *dir;
	const char *filename;
	char *pathname, *buf;
	int fd, bytes;
	svector *results = new svector();
	
	// Access /proc/<number>/cmdline
	pathname = new char[80];
	buf = new char[500];
	dir = opendir("/proc");
	if(dir == NULL)
	{
		warning("Cannot open /proc directory; disabling running process check");
		delete[] pathname;
		delete[] buf;
		return results;
	}
	while(1)
	{
		de = readdir(dir);
		if(de == NULL)
			break;
		filename = de->d_name;
		if(!is_an_integer(filename))
			continue;
		sprintf(pathname, "/proc/%s/cmdline", filename);
		fd = open(pathname, O_RDONLY);
		if(fd == -1)
		{
			printf("Warning: could not open %s\n", pathname);
			continue;
		}
		bytes = read(fd, buf, 500);
		close(fd);
		if(buf[bytes - 1] != '\0')
			continue;
		for(int i = 0; i < bytes - 1; i++)
		{
			if(buf[i] == '\0')
				buf[i] = ' ';
		}
		results->add(buf);
	}
	closedir(dir);
	delete[] pathname;
	delete[] buf;
	return results;
}
